### ISSUE TITLE
Meson build must use igr-runner for integration tests

### ISSUE THREAD

**Issue by davmac314** [davmac314/dinit#302]
The integration tests are to be re-worked to avoid over-reliance on shell scripts. This has begun, with commit 49a9da28.

The scripts for the tests are still in place for now but will need to be removed. The meson build must be altered to account for this.

**Comment by mobin-2008** [comment]
My plan for this is to seprated in two section:

1. Make or Meson which have targets for each test.

2. `igr_runner` which contains actual tests and runs passed test, A example of invoked igr_runner: `./igr_runner basic`

My main reason for that is parallelism, In Make users can use `-j` for parallel run and in Meson it's default.

Also if list would be in Makefile, We probably can easily read it from Meson side and it makes into less trouble for me specifically in case of Meson list of tests.

I'm going to go deeper into changes you've made.

**Comment by davmac314** [comment]
> Also if list would be in Makefile, We probably can easily read it from Meson side

Maybe have `igr-runner` output a list of tests if given a particular command line argument (`--list-tests` or similar). I don't want the Meson build trying to read the makefile(s), that it is going to be fragile and error-prone.

Also for now at least please leave the integration tests as a single target in the makefiles. This has benefits such as that `igr-runner` can produce a summary of the test results. The parallelism isn't important at this stage (the tests don't take long to run) and it can be added later anyway (if necessary) by making igr-runner support the jobserver protocol.

**Comment by davmac314** [comment]
> Hmm, Looks like we couldn't use --list-tests in meson because building section is splited from configuration section, so it should be a static list I guess.

That's a shame. Really feels like Meson is quite limited.

**Comment by mobin-2008** [comment]
@davmac314 Are you going to implement `igr-runner test_name` syntax yourself? Looks like current code lacks command-line flags proccesing.

**Comment by davmac314** [comment]
I might but it is not high priority for me. It is not needed for the makefile build.

**Comment by mobin-2008** [comment]
> I might but it is not high priority for me. It is not needed for the makefile build.

Anyway I will fix it after #263

**Comment by davmac314** [comment]
This needs to take high priority, if we want meson support in 1.0 release.

**Comment by davmac314** [comment]
Two possibilities that I can think of for handling this are:

1. run all the igr_runner tests as if they were a single test in meson
2. implement extraction of the test names from igr_runner, and do this as a release step to create a json (or similar) file that can then be used by the meson build. So the list of tests becomes part of the release tarball. One downside is building from git will need an extra step.

There might be other possibilities.

**Comment by mobin-2008** [comment]
There is a lot of different issues about igr-runner, such as some race conditions, Some `XXX` around variable `DINIT_BINDIR` and more. Probably fixing all of issues will take a long time.

@davmac314 When you want to release next version?

**Comment by davmac314** [comment]
Next release will probably be 1.0 but I can't say when. The remaining integration tests need to be converted to run under igr-runner, FreeBSD boot needs to be sorted, and I would like to add some mechanism for user (admin) configurability of system (distribution) services.

But this issue should be high priority. Any new tests will not work except via igr-runner.

> There is a lot of different issues about igr-runner, such as some race conditions

What race conditions?

>  Some XXX around variable DINIT_BINDIR and more. Probably fixing all of issues will take a long time.

I do not think so.

**Comment by davmac314** [comment]
By the way the `XXX` comments were to specifically mark things that might need some extra logic for the meson build. It should be easy enough to add command line arguments to `igr-runner` in order to set `dinit_bindir` and `igr_output_basedir`.

**Comment by mobin-2008** [comment]
I fixed them up, the race was about same path for all sockets. It's ready to be merged but I waiting for you to completely port all tests to new igr-runner.

**Comment by davmac314** [comment]
The integration tests are all running via igr-runner now.
